Mark cache directories as excluded from backup

szewai opened the pull request:
#### d84c620e7679fd8b8b649d7a67e4dacace3a216a
<pre>
Mark cache directories as excluded from backup
<a href="https://bugs.webkit.org/show_bug.cgi?id=247703">https://bugs.webkit.org/show_bug.cgi?id=247703</a>
rdar://102164410

Reviewed by Geoffrey Garen.

These directories were excluded from backup because they were under Caches/ directory by default. After 255271@main, the
directories of a WebsiteDataStore created with UUID can be outside of Caches/ directory, so we need to explicitly mark
them excluded to keep the old behavior.

* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::setExcludedFromBackup):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::resolveDirectoriesIfNecessary):

Canonical link: <a href="https://commits.webkit.org/256579@main">https://commits.webkit.org/256579@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/49209a7abf6f0be43ded03ab0b00221c515467c3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96144 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5394 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105689 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166030 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5516 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34156 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88521 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101511 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101803 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4089 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82749 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31110 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87848 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73945 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/87154 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39884 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/82521 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37558 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20720 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28250 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4559 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42154 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43324 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/85200 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/44045 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39980 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19219 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -339,6 +339,14 @@ void WebsiteDataStore::resolveDirectoriesIfNecessary()
         m_resolvedConfiguration->setCookieStorageFile(FileSystem::pathByAppendingComponent(resolvedCookieDirectory, FileSystem::pathFileName(m_configuration->cookieStorageFile())));
     }
 
+    // Do not back up cache type data.
```
You can write out your comment in code like this:

```
Vector<String> allCacheDirectories = {
    resolvedApplicationCacheDirectory(),
    resolvedMediaCacheDirectory(),
    resolvedNetworkCacheDirectory(),
#if ENABLE(ARKIT_INLINE_PREVIEW)
    resolvedModelElementCacheDirectory(),
#endif
};

for (const auto& directory : allCacheDirectories) {
    FileSystem::setExcludedFromBackup(directory, true);
};
```

Future diffs that add or remove cache directories get simpler this way too.

cdumez replied:
Do we really need a Vector though, wouldn't an array suffice?
```
std::array allCacheDirectories = {
    resolvedApplicationCacheDirectory(),
    resolvedMediaCacheDirectory(),
    resolvedNetworkCacheDirectory(),
#if ENABLE(ARKIT_INLINE_PREVIEW)
    resolvedModelElementCacheDirectory(),
#endif
};
```

geoffreygaren replied:
That's good too! I think std::array might still require that you specify the count as a template parameter.

cdumez replied:
> I think std::array might still require that you specify the count as a template parameter.
Not with C++17 AFAIK.

cdumez replied:
https://en.cppreference.com/w/cpp/container/array/deduction_guides

geoffreygaren reviewed (approved):
r=me

szewai committed:
Mark cache directories as excluded from backup
https://bugs.webkit.org/show_bug.cgi?id=247703
rdar://102164410

Reviewed by Geoffrey Garen.

These directories were excluded from backup because they were under Caches/ directory by default. After 255271@main, the
directories of a WebsiteDataStore created with UUID can be outside of Caches/ directory, so we need to explicitly mark
them excluded to keep the old behavior.

* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::setExcludedFromBackup):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::resolveDirectoriesIfNecessary):

Canonical link: https://commits.webkit.org/256579@main
